packages/fast-components-react-base/src/viewport-positioner/viewport-positioner.tsx

```diff
@@ -0,0 +1,1011 @@
+            desiredHorizontalPosition = this.state.defaultHorizontalPosition;
+
+            const horizontalThreshold: number =
+                this.props.horizontalThreshold !== undefined
+                    ? this.props.horizontalThreshold
+                    : this.positionerRect.width;
+
+            if (
+                desiredHorizontalPosition ===
+                    ViewportPositionerHorizontalPositionLabel.undefined ||
+                this.getOptionWidth(desiredHorizontalPosition) < horizontalThreshold
+            ) {
+                desiredHorizontalPosition =
+                    this.getOptionWidth(horizontalOptions[0]) >
+                    this.getOptionWidth(horizontalOptions[1])
+                        ? horizontalOptions[0]
+                        : horizontalOptions[1];
+            }
+        }
+
+        if (this.props.verticalPositioningMode !== AxisPositioningMode.uncontrolled) {
+            const verticalOptions: ViewportPositionerVerticalPositionLabel[] = this.getVerticalPositioningOptions();
+            desiredVerticalPosition = this.state.defaultVerticalPosition;
+
+            const verticalThreshold: number =
+                this.props.verticalThreshold !== undefined
+                    ? this.props.verticalThreshold
+                    : this.positionerRect.height;
+
+            if (
+                desiredVerticalPosition ===
+                    ViewportPositionerVerticalPositionLabel.undefined ||
+                this.getOptionHeight(desiredVerticalPosition) < verticalThreshold
+            ) {
+                desiredVerticalPosition =
+                    this.getOptionHeight(verticalOptions[0]) >
+                    this.getOptionHeight(verticalOptions[1])
+                        ? verticalOptions[0]
+                        : verticalOptions[1];
+            }
+        }
+
+        this.setState(Object.assign(
+            {
+                xTranslate: this.calculateHorizontalTranslate(desiredHorizontalPosition),
+                yTranslate: this.calculateVerticalTranslate(desiredVerticalPosition),
+                initialLayoutComplete: true,
+            },
+            this.getBaseHorizontalPositioningState(desiredHorizontalPosition),
+            this.getBaseVerticalPositioningState(desiredVerticalPosition)
+        ) as ViewportPositionerState);
+    };
+
+    /**
+     * Get base horizontal positioning state based on desired position
+     */
+    private getBaseHorizontalPositioningState = (
+        desiredHorizontalPosition: ViewportPositionerHorizontalPositionLabel
+    ): Partial<ViewportPositionerState> => {
+        let right: number = null;
+        let left: number = null;
+        let xTransformOrigin: string = "left";
+
+        switch (desiredHorizontalPosition) {
+            case ViewportPositionerHorizontalPositionLabel.left:
+                xTransformOrigin = "right";
```

Comment:
could you use an enum for these strings?

Reply:
yep